File: cpp/lib/token.cpp
#include "monkey/token.h"

#include <fmt/ostream.h>

#include <unordered_map>
#include <utility>

using fmt::format;
using std::string;
using std::unordered_map;

namespace monkey {

string type_to_str(Token::Type type) {
  static unordered_map<Token::Type, string> lookup{
      {Token::Type::ILLEGAL, "ILLEGAL"},
      {Token::Type::EOF_, "EOF"},
      {Token::Type::IDENT, "IDENT"},
      {Token::Type::INT, "INT"},
      {Token::Type::ASSIGN, "ASSIGN"},
      {Token::Type::PLUS, "PLUS"},
      {Token::Type::MINUS, "MINUS"},
      {Token::Type::BANG, "BANG"},
      {Token::Type::ASTERISK, "ASTERISK"},
      {Token::Type::SLASH, "SLASH"},
      {Token::Type::GT, "GT"},
      {Token::Type::LT, "LT"},
      {Token::Type::EQ, "EQ"},
      {Token::Type::NOT_EQ, "NOT_EQ"},
      {Token::Type::COMMA, "COMMA"},
      {Token::Type::SEMICOLON, "SEMICOLON"},
      {Token::Type::LPAREN, "LPAREN"},
      {Token::Type::RPAREN, "RPAREN"},
      {Token::Type::LBRACE, "LBRACE"},
      {Token::Type::RBRACE, "RBRACE"},
      {Token::Type::FUNCTION, "FUNCTION"},
      {Token::Type::LET, "LET"},
      {Token::Type::TRUE, "TRUE"},
      {Token::Type::FALSE, "FALSE"},
      {Token::Type::IF, "IF"},
      {Token::Type::ELSE, "ELSE"},
      {Token::Type::RETURN, "RETURN"},
  };
  return lookup.find(type)->second;
}

Token::Type Token::lookup_ident(const std::string& ident) {
  static unordered_map<string, Token::Type> keywords{
      {"fn", Token::Type::FUNCTION},
      {"let", Token::Type::LET},
      {"true", Token::Type::TRUE},
      {"false", Token::Type::FALSE},
      {"if", Token::Type::IF},
      {"else", Token::Type::ELSE},
      {"return", Token::Type::RETURN},
  };

  auto s = keywords.find(ident);
  if (s == keywords.end()) return Type::IDENT;
  return s->second;
}

bool Token::operator==(const Token& other) const {
  return type == other.type && literal == other.literal;
}

ostream& operator<<(ostream& os, const Token& tok) {
  return os << format("{}<\"{}\">", tok.type, tok.literal);
}

ostream& operator<<(ostream& os, Token::Type type) {
  return os << type_to_str(type);
}
Token::Token() = default;
Token::Token(Token::Type type, string literal)
    : type{type}
    , literal{move(literal)} { }
Token::Token(const Token&) = default;
Token::Token(Token&&) noexcept = default;
Token& Token::operator=(Token&&) noexcept = default;
} // namespace monkey

File: rs/src/token.rs
use phf::{phf_map};

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum TokenType {
    Illegal,
    // EOF,

    Ident,
    Int,
    Str,

    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,

    LT,
    GT,
    Eq,
    NotEq,

    Comma,
    Semicolon,
    Colon,

    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Function,
    Let,
    True,
    False,
    If,
    Else,
    Return,
}

#[derive(Debug, PartialEq)]
pub struct Token {
    pub type_: TokenType,
    pub literal: String,
    pub row: usize,
    pub col: usize,
}

impl Token {
    pub fn new() -> Token {
        Token {
            row: 0,
            col: 0,
            type_: TokenType::Illegal,
            literal: String::from(""),
        }
    }
}

static KEYWORDS: phf::Map<&'static str, TokenType> = phf_map! {
    "fn" => TokenType::Function,
    "let" => TokenType::Let,
    "return" => TokenType::Return,
    "if" => TokenType::If,
    "else" => TokenType::Else,
    "true" => TokenType::True,
    "false" => TokenType::False,
};

pub fn lookup_ident(ident: &String) -> TokenType {
    match KEYWORDS.get(ident as &str) {
        Some(t) => *t,
        _ => TokenType::Ident,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lookup_ident() {
        let id = lookup_ident(&String::from("xxx"));
        assert_eq!(id, TokenType::Ident);

        let id = lookup_ident(&String::from("fn"));
        assert_eq!(id, TokenType::Function);
    }
}
